### PR TITLE
✨ AoG: Add amp-viewer-assistance extension.

### PR DESCRIPTION
**Pull request by hellokoji**
Implements the amp-viewer-assistance extension proposed in Issue (#15195).

The new extension creates a messaging channel between AMP pages and AMP viewers to optimize for flows between the embedded page and the external viewer.

cc @choumx

### CODE REVIEW

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#
+# Copyright 2016 The AMP HTML Authors. All Rights Reserved.
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: s/2016/2019 in all files

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+tags: {  # amp-viewer-assistance
+  html_format: AMP
+  html_format: EXPERIMENTAL
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: s/EXPERIMENTAL/ACTIONS

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    version: "0.1"
+    version: "latest"
+    requires_usage: GRANDFATHERED
+    deprecated_allow_duplicates: true
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {  # amp-viewer-assistance (json)
+  html_format: AMP
+  html_format: EXPERIMENTAL
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: s/EXPERIMENTAL/ACTIONS

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#
+tags: {  # amp-viewer-assistance
+  html_format: AMP
+  html_format: EXPERIMENTAL
+  tag_name: "SCRIPT"
+  extension_spec: {
+    name: "amp-viewer-assistance"
+    version: "0.1"
+    version: "latest"
+    requires_usage: GRANDFATHERED
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Remove

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    deprecated_allow_duplicates: true
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Remove

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 1 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Done.

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 2 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Done.

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 3 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Done.

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 5 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Done.

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 4 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Done.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  /**
+   * Toggles the CSS classes related to the status of the identity token.
+   * @private
+   * @param {boolean} available
+   */
+  setIdTokenStatus_(available) {
+    this.toggleTopClass_(
+        'amp-viewer-assistance-identity-available', available);
+    this.toggleTopClass_(
+        'amp-viewer-assistance-identity-unavailable', !available);
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Why do we need two classes for the same boolean state?

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  /**
+   * @return {!Promise<undefined>}
+   */
+  getIdTokenPromise() {
+    return this.viewer_.sendMessageAwaitResponse('getAccessTokenPassive', dict({
+      // For now there's only 1 provider option, so we just hard code it
+      'providers': [GSI_TOKEN_PROVIDER],
+    }))
+        .then(token => {
+          this.setIdTokenStatus_(Boolean(token));
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: `!!token`

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  /**
+   * @private
+   * @restricted
+   */
+  start_() {
+    if (!this.enabled_) {
+      user().info(
+          TAG, 'Invalid AMP Action - no "id=amp-viewer-assistance" element');
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: `user().error(TAG, 'Could not find #amp-viewer-assistance element.`);`

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    } else if (method == 'signIn') {
+      this.requestSignIn_();
+    }
+
+    return null;
+  }
+
+  /**
+   * @private
+   * @restricted
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: `@return {!AmpViewerAssistance}`

File: extensions/amp-viewer-assistance/amp-viewer-assistance.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# amp-viewer-assistance
+
+The amp-viewer-assistance element provides specification of AMP viewer
+configuration information and invocation of AMP viewer assisted behavior.
```
Comment (choumx) [inline extensions/amp-viewer-assistance/amp-viewer-assistance.md]: This documentation needs some love. :) 

We ought to:
1. Clearly describe the high-level product use case. E.g. this is meant for "assistive behaviors on AMP pages, facilitated by the viewer."
2. Document AMP page/viewer API contract e.g. "requestSignIn". 
3. Document element actions like "signIn" and "updateActionState". 
4. Follow the format from other extension documentation (see amp-access.md).

/cc @CrystalOnScript

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+  requestSignIn_() {
+    this.viewer_.sendMessageAwaitResponse('requestSignIn', dict({
+      'providers': [GSI_TOKEN_PROVIDER],
+    })).then(token => {
+      user().info(TAG, 'Token: ' + token);
+      if (token) {
+        this.setIdTokenStatus_(/*available=*/true);
+        this.action_.trigger(
+            this.assistanceElement_, 'signedIn', null, ActionTrust.HIGH);
+      }
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Do we intentionally not set token status to unavailable on sign-in failure?

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+   */
+  start_() {
+    if (!this.enabled_) {
+      user().info(
+          TAG, 'Invalid AMP Action - no "id=amp-viewer-assistance" element');
+      return this;
+    }
+    return this.viewer_.isTrustedViewer().then(isTrustedViewer => {
+      if (!isTrustedViewer &&
+         !isExperimentOn(this.ampdoc_.win, 'amp-viewer-assistance-untrusted')) {
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Is this intended to be enabled eventually?

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        this.enabled_ = false;
+        user().info(TAG, 'Disabling AMP Action since viewer is not trusted');
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: `user().error(TAG, "amp-viewer-assistance is currently only supported on trusted viewers.");`

File: src/url.js [hunk 14]
```diff
@@ -36,6 +36,8 @@ const SERVING_TYPE_PREFIX = dict({
   'a': true,
   // Ad
   'ad': true,
+  // Actions viewer
+  'action': true,
```
Comment (choumx) [inline src/url.js]: Why is this necessary? I don't see any usages of url.js functions in amp-viewer-assistance.js.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 6 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: It was originally to provide 3 states. One for the identity token being available, one for unavailable, and one for the viewer not being initiated (no css class). We ended up not using the three states, so I removed the unavailable class.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 7 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Done.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 8 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Done.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 9 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Done.

File: extensions/amp-viewer-assistance/amp-viewer-assistance.md (hunk 10 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/amp-viewer-assistance.md]: Agreed! More documentation will be committed to this PR soon.

P.S. That is a very nice way of requesting for more documentation! :+1: :smile:

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 11 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: The token status is set to unavailable by default and then switched on in this request. Since there is not a way to have a token be available => unavailable, there was never a need for the failure case. Regardless, adding in the other case to be thorough.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 12 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: No, the experiment was our method of forgoing the trusted viewer check with testing locally since local viewers are inherently not trusted.

If there are other ways of accomplishing the same effect, we are open to suggestions.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 13 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Done.

File: src/url.js (hunk 14 above)
Comment (hellokoji) [reply src/url.js]: You're right that we do not use `url.js` in `amp-viewer-assistance.js`. However 'action' was added to the  in order to whitelist 'action' as a valid prefix for [this assertion](https://github.com/ampproject/amphtml/blob/3c134ac06763ba485c8bf96c71799d96be563aca/src/url.js#L510).

File: extensions/amp-viewer-assistance/amp-viewer-assistance.md (hunk 10 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/amp-viewer-assistance.md]: Just updated the documentation to address some of those comments.

**Comment by hellokoji** [comment]
Friendly ping.

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii [hunk 15]
```diff
@@ -0,0 +1,53 @@
+#
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: This file already exists as [validator-amp-viewer-assistance.protoascii](https://github.com/ampproject/amphtml/blob/master/extensions/amp-viewer-assistance/validator-amp-viewer-assistance.protoascii)

File: extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.html [hunk 16]
```diff
@@ -0,0 +1,38 @@
+<!--
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.html]: This file already exists as [validator-actions-amp-viewer-assistance.html](https://github.com/ampproject/amphtml/blob/master/extensions/amp-viewer-assistance/0.1/test/validator-actions-amp-viewer-assistance.html)

File: extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.out [hunk 17]
```diff
@@ -0,0 +1,39 @@
+PASS
```
Comment (honeybadgerdontcare) [inline extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.out]: This file already exists as [validator-actions-amp-viewer-assistance.out](https://github.com/ampproject/amphtml/blob/master/extensions/amp-viewer-assistance/0.1/test/validator-actions-amp-viewer-assistance.out)

File: extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii (hunk 15 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/validator-viewer-assistance.protoascii]: Removed.

File: extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.html (hunk 16 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.html]: Removed.

File: extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.out (hunk 17 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/test/validator-amp-viewer-assistance.out]: Removed.

File: src/url.js (hunk 14 above)
Comment (choumx) [reply src/url.js]: Can we extract this to a separate PR? I'm also curious what the specific codepath (which caller of `getSourceUrl()`) triggers that assertion for you.

File: tools/experiments/experiments.js [hunk 18]
```diff
@@ -346,6 +346,10 @@ const EXPERIMENTS = [
     name: 'Enables scroll snap on carousel on Chrome browsers',
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/16508',
   },
+  {
+    id: 'amp-viewer-assistance-untrusted',
+    name: 'Enable amp-viewer-assistance in untrusted viewers',
+  },
```
Comment (choumx) [inline tools/experiments/experiments.js]: Can we remove this? It's a bit confusing for a local-testing trigger to be listed with other experiments. Or do you need this for local testing on device?

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js (hunk 12 above)
Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Yes, this is specifically for developing/testing with a local viewer using a remote (or local) runtime. Does that make sense?

File: src/url.js (hunk 14 above)
Comment (hellokoji) [reply src/url.js]: We can move this to a different PR, however it would be a breaking change. The codepath is whenever `getSourceUrl()` is called, most likely during [the open channel](https://github.com/ampproject/amphtml/blob/990fc5961efa6c5ed8affa7fd176ffa340277275/extensions/amp-viewer-integration/0.1/amp-viewer-integration.js#L170) call in `amp-viewer-integration.js`. It's necessary because the cdn url requests have an `action` prefix present.

File: tools/experiments/experiments.js (hunk 18 above)
Comment (hellokoji) [reply tools/experiments/experiments.js]: Yes, removed.

File: extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    /** @private @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(ampdoc.win);
+  }
+
+  /**
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {?Promise}
+   * @private
+   */
+  actionHandler_(invocation) {
```
Comment (choumx) [inline extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: Sorry didn't catch this earlier. You should actually use `this.registerAction()` here instead. See `base-element.js#registerAction`.

Then you can also remove the whitelisting in `presubmit-checks.js`.

Comment (hellokoji) [reply extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js]: As we discussed offline, we are following the pattern set by [amp-access](https://github.com/ampproject/amphtml/blob/b1993f31d96bd7c287f7a7e47b8c9e2534def2d8/extensions/amp-access/0.1/amp-access.js#L630) to justify using our own action handler since we are using a `<script>` tag instead of an `<amp-viewer-assistance>` element.
